src/RBGS.f

```diff
@@ -938,4 +938,955 @@ subroutine rbgs_mask(
 c
       end subroutine
 
+cccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccc
+c
+c       Computes sol to a linear system on each cell in a patch i.e. box relaxation
+c
+c       where un, us are vector valued side-centered velocities
+c             p is a scalar-valued cell-centered pressure
+c       and   thn is a scalar-valued cell-centered network volume fraction
+c
+cccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccc
+      subroutine rbgs_var_xi(
+     &        dx, ilow0, iup0,
+     &        ilow1, iup1,
+     &        un_data_0, un_data_1, un_gcw,
+     &        us_data_0, us_data_1, us_gcw,
+     &        p_data, p_gcw, f_p_data, f_p_gcw,
+     &        f_un_data_0, f_un_data_1, f_un_gcw,
+     &        f_us_data_0, f_us_data_1, f_us_gcw,
+     &        thn_data, thn_gcw, eta_n, eta_s,
+     &        xi_0, xi_1, xi_gcw,
+     &        w, C, D, red_or_black)
+c
+      use my_subs
+      implicit none
+cccccccccccccccccccccccccccccccccc INPUTS ccccccccccccccccccccccccccccc
+      double precision dx(0:1)
+      integer ilow0,  iup0
+      integer ilow1,  iup1
+      integer un_gcw, us_gcw, p_gcw, f_un_gcw, f_us_gcw
+      integer f_p_gcw, thn_gcw, red_or_black
+      double precision eta_n, eta_s, w, C, D
+c
+      double precision thn_data(ilow0-thn_gcw:iup0+thn_gcw,
+     &          ilow1-thn_gcw:iup1+thn_gcw)
+c
+      double precision un_data_0(ilow0-un_gcw:iup0+un_gcw+1,
+     &          ilow1-un_gcw:iup1+un_gcw)
+      double precision un_data_1(ilow0-un_gcw:iup0+un_gcw,
+     &          ilow1-un_gcw:iup1+un_gcw+1)
+c
+      double precision us_data_0(ilow0-us_gcw:iup0+us_gcw+1,
+     &          ilow1-us_gcw:iup1+us_gcw)
+      double precision us_data_1(ilow0-us_gcw:iup0+us_gcw,
+     &          ilow1-us_gcw:iup1+us_gcw+1)
+c
+      double precision p_data(ilow0-p_gcw:iup0+p_gcw,
+     &          ilow1-p_gcw:iup1+p_gcw)
+
+      double precision f_un_data_0(ilow0-f_un_gcw:iup0+f_un_gcw+1,
+     &          ilow1-f_un_gcw:iup1+f_un_gcw)
+      double precision f_un_data_1(ilow0-f_un_gcw:iup0+f_un_gcw,
+     &          ilow1-f_un_gcw:iup1+f_un_gcw+1)
+c
+      double precision f_us_data_0(ilow0-f_us_gcw:iup0+f_us_gcw+1,
+     &          ilow1-f_us_gcw:iup1+f_us_gcw)
+      double precision f_us_data_1(ilow0-f_us_gcw:iup0+f_us_gcw,
+     &          ilow1-f_us_gcw:iup1+f_us_gcw+1)
+c
+      double precision f_p_data(ilow0-f_p_gcw:iup0+f_p_gcw,
+     &          ilow1-f_p_gcw:iup1+f_p_gcw)
+
+      integer xi_gcw
+      double precision xi_0(ilow0-xi_gcw:iup0+xi_gcw+1,
+     &                      ilow1-xi_gcw:iup1+xi_gcw)
+      double precision xi_1(ilow0-xi_gcw:iup0+xi_gcw,
+     &                      ilow1-xi_gcw:iup1+xi_gcw+1)
+c
```

Comment:
Similar as above, I wrote multiple functions to do basically the same thing because these are efficiency sensitive routines. If we pass in variable drag coefficient data, that takes up 2 * N * (N+1)*8 bytes of data, which for my processor, consists of 1/4 of it's L1 cache for a 64 by 64 patch. If we pass a single number, that's 8 bytes.

Now this is probably prematurely optimizing, but I don't have the time to fully check the efficiency of this code. So my reasoning was to duplicate the code and let the compiler handle it.